smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,221 @@
+
+    private WebsocketConnectionAttemptState connectionAttemptState;
+
+    protected XmppWebsocketTransportModule(XmppWebsocketTransportModuleDescriptor moduleDescriptor,
+                    ModularXmppClientToServerConnectionInternal connectionInternal) {
+        super(moduleDescriptor, connectionInternal);
+
+        websocketTransport = new XmppWebsocketTransport(connectionInternal);
+    }
+
+    private void afterOutgoingElementsQueueModified() {
+        TopLevelStreamElement topLevelStreamElement = connectionInternal.outgoingElementsQueue.poll();
+        connectionAttemptState.websocket.send(topLevelStreamElement);
+    }
+
+    @Override
+    protected XmppWebsocketTransport getTransport() {
+        return websocketTransport;
+    }
+
+    static final class EstablishingWebsocketConnectionStateDescriptor extends StateDescriptor {
+        private EstablishingWebsocketConnectionStateDescriptor() {
+            super(XmppWebsocketTransportModule.EstablishingWebsocketConnectionState.class);
+            addPredeccessor(LookupRemoteConnectionEndpointsStateDescriptor.class);
+            addSuccessor(ConnectedButUnauthenticatedStateDescriptor.class);
+            declareInferiortyTo(EstablishingTcpConnectionStateDescriptor.class);
+        }
+
+        @Override
+        protected State constructState(ModularXmppClientToServerConnectionInternal connectionInternal) {
+            XmppWebsocketTransportModule websocketTransportModule = connectionInternal.connection.getConnectionModuleFor(
+                            XmppWebsocketTransportModuleDescriptor.class);
+            return websocketTransportModule.constructEstablishingWebsocketConnectionState(this, connectionInternal);
+        }
+    }
+
+    final class EstablishingWebsocketConnectionState extends State {
+        protected EstablishingWebsocketConnectionState(StateDescriptor stateDescriptor,
+                        ModularXmppClientToServerConnectionInternal connectionInternal) {
+            super(stateDescriptor, connectionInternal);
+        }
+
+        @Override
+        public AttemptResult transitionInto(WalkStateGraphContext walkStateGraphContext)
+                        throws IOException, SmackException, InterruptedException, XMPPException {
+
+            connectionAttemptState = new WebsocketConnectionAttemptState(
+                            connectionInternal, discoveredWebsocketEndpoints, this);
+
+            StateTransitionResult.Failure failure = connectionAttemptState.establishWebsocketConnection();
+            if (failure != null) {
+                return failure;
+            }
+
+            connectionInternal.setTransport(websocketTransport);
+
+            connectionInternal.newStreamOpenWaitForFeaturesSequence("stream features after initial connection",
+                    new OpenElement(connectionInternal.connection.getXMPPServiceDomain()));
+
+            return new WebsocketConnectedResult();
```

Comment:
Just like the `EstablishTcpConnectionState` this result should contain dynamic information about the established connection. For example, the used remote endpoint / address.